File: code/ahasmove/105.从前序与中序遍历序列构造二叉树.cpp
/*
 * @lc app=leetcode.cn id=105 lang=cpp
 *
 * [105] 从前序与中序遍历序列构造二叉树
 *
 * https://leetcode-cn.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/description/
 *
 * algorithms
 * Medium (68.54%)
 * Likes:    764
 * Dislikes: 0
 * Total Accepted:    129.9K
 * Total Submissions: 189.5K
 * Testcase Example:  '[3,9,20,15,7]\n[9,3,15,20,7]'
 *
 * 根据一棵树的前序遍历与中序遍历构造二叉树。
 * 
 * 注意:
 * 你可以假设树中没有重复的元素。
 * 
 * 例如，给出
 * 
 * 前序遍历 preorder = [3,9,20,15,7]
 * 中序遍历 inorder = [9,3,15,20,7]
 * 
 * 返回如下的二叉树：
 * 
 * ⁠   3
 * ⁠  / \
 * ⁠ 9  20
 * ⁠   /  \
 * ⁠  15   7
 * 
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

#include <vector>
using namespace std;

class Solution
{
public:
    TreeNode *divdeCounter(vector<int> &preorder, int l1, int r1, vector<int> &inorder, int l2, int r2)
    {

        if (r1 < l1)
            return nullptr;

        int midVal = preorder[l1];
        int mid = l2;
        while (midVal != inorder[mid])
            mid++;

        TreeNode *node = new TreeNode(midVal);

        node->left = divdeCounter(preorder, l1 + 1, l1 + mid - l2, inorder, l2, mid - 1);
        node->right = divdeCounter(preorder, l1 + mid - l2 + 1, r1, inorder, mid + 1, r2);
        // 1 3 5 6
        // 3 1 5 6
        return node;
    }

    TreeNode *buildTree(vector<int> &preorder, vector<int> &inorder)
    {
        return divdeCounter(preorder, 0, preorder.size() - 1,
                            inorder, 0, inorder.size() - 1);
    }
};
// @lc code=end

File: code/JavaLab/src/main/java/algorithm/leetcode/tree/105.从前序与中序遍历序列构造二叉树.java
package algorithm.leetcode.tree;

import algorithm.config.TreeNode;

/**
 * https://leetcode-cn.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/
 * <p>
 * 给定一棵树的前序遍历 preorder 与中序遍历  inorder。请构造二叉树并返回其根节点。
 */
class LC105 {

    public TreeNode buildTree(int[] preorder, int[] inorder) {
        return helper(preorder, 0, preorder.length - 1, inorder, 0, inorder.length - 1);
    }


    /**
     * @param preorder 前序遍历顺序
     * @param l1       当前子树前序遍历起点
     * @param r1       当前子树前序遍历终点
     * @param inorder  中序遍历顺序
     * @param l2       当前子树中序遍历起点
     * @param r2       当前子树中序遍历终点
     * @return
     */
    public TreeNode helper(int[] preorder, int l1, int r1, int[] inorder, int l2, int r2) {
        if (l1 > r1) {
            return null;
        }

        // 当前子树跟节点(先序遍历首个节点)
        int rootVal = preorder[l1];
        TreeNode root = new TreeNode(rootVal);
        // 从中序遍历中找到根节点
        int index = l2;
        while (inorder[index] != rootVal)
            index++;
        int diff = index - l2;

        // 根节点左侧为左子树序列，右侧为右子树序列
        root.left = helper(preorder, l1 + 1, l1 + diff, inorder, l2, index - 1);
        root.right = helper(preorder, l1 + diff + 1, r1, inorder, index + 1, r2);
        return root;
    }
}